daemon/nvidia_linux.go

```diff
@@ -38,70 +40,96 @@ func init() {
 		capset:     capset,
 		updateSpec: setNvidiaGPUs,
 	}
-	for c := range capset {
-		nvidiaDriver.capset[c] = struct{}{}
+	for c := range allNvidiaCaps {
+		nvidiaDriver.capset[string(c)] = struct{}{}
 	}
 	registerDeviceDriver("nvidia", nvidiaDriver)
 }
 
 func setNvidiaGPUs(s *specs.Spec, dev *deviceInstance) error {
-	var opts []nvidia.Opts
+	envMap, err := getEnvMap(s.Process.Env)
```

Comment:
The conversion to a map is only to make sure we "overwrite" the previous value, correct?

Given that we're only appending new values (and don't look at existing values), I don't think there's a need to do this conversion, as `FOOBAR=one FOOBAR=two` is equivalent to `FOOBAR=two` (latter value overwrites former value);

```patch
diff --git a/daemon/nvidia_linux.go b/daemon/nvidia_linux.go
index 6756187649..86e4c0d006 100644
--- a/daemon/nvidia_linux.go
+++ b/daemon/nvidia_linux.go
@@ -1,7 +1,6 @@
 package daemon
 
 import (
-       "fmt"
        "os"
        "os/exec"
        "strconv"
@@ -47,22 +46,17 @@ func init() {
 }
 
 func setNvidiaGPUs(s *specs.Spec, dev *deviceInstance) error {
-       envMap, err := getEnvMap(s.Process.Env)
-       if err != nil {
-               return err
-       }
-
        req := dev.req
        if req.Count != 0 && len(req.DeviceIDs) > 0 {
                return errConflictCountDeviceIDs
        }
 
        if len(req.DeviceIDs) > 0 {
-               envMap["NVIDIA_VISIBLE_DEVICES"] = strings.Join(req.DeviceIDs, ",")
+               s.Process.Env = append(s.Process.Env, "NVIDIA_VISIBLE_DEVICES="+strings.Join(req.DeviceIDs, ","))
        } else if req.Count > 0 {
-               envMap["NVIDIA_VISIBLE_DEVICES"] = countToDevices(req.Count)
+               s.Process.Env = append(s.Process.Env, "NVIDIA_VISIBLE_DEVICES="+countToDevices(req.Count))
        } else if req.Count < 0 {
-               envMap["NVIDIA_VISIBLE_DEVICES"] = "all"
+               s.Process.Env = append(s.Process.Env, "NVIDIA_VISIBLE_DEVICES=all")
        }
 
        var nvidiaCaps []string
@@ -78,11 +72,9 @@ func setNvidiaGPUs(s *specs.Spec, dev *deviceInstance) error {
        }
 
        if nvidiaCaps != nil {
-               envMap["NVIDIA_DRIVER_CAPABILITIES"] = strings.Join(nvidiaCaps, ",")
+               s.Process.Env = append(s.Process.Env, strings.Join(nvidiaCaps, ","))
        }
 
-       s.Process.Env = envMapToArray(envMap)
-
        path, err := exec.LookPath(nvidiaHook)
        if err != nil {
                return err
@@ -111,25 +103,3 @@ func countToDevices(count int) string {
        }
        return strings.Join(devices, ",")
 }
-
-func getEnvMap(e []string) (map[string]string, error) {
-       m := make(map[string]string)
-       for _, s := range e {
-               p := strings.SplitN(s, "=", 2)
-               if len(p) != 2 {
-                       return nil, fmt.Errorf("Failed to split environment variable `%s`", p)
-               }
-               m[p[0]] = p[1]
-       }
-       return m, nil
-}
-
-func envMapToArray(m map[string]string) []string {
-       env := make([]string, len(m))
-       i := 0
-       for k, v := range m {
-               env[i] = fmt.Sprintf("%s=%s", k, v)
-               i++
-       }
-       return env
-}
```

Reply:
This is a lot better!

Thanks!